commons/src/main/java/esa/commons/io/AbstractPathWatcher.java

```diff
@@ -0,0 +1,265 @@
+                watch();
+            }
+        });
+    }
+
+    @Override
+    public boolean stopAndWait(long timeout, TimeUnit unit) throws InterruptedException {
+        if (stopped) {
+            if (delayScheduler == null) {
+                return true;
+            } else {
+                return delayScheduler.awaitTermination(timeout, unit);
+            }
+        }
+        LOGGER.info("PathWatcher of {} is stopping!", path);
+
+        //Set stop to true firstly , so that the watch() method will be terminated at the
+        //first time when the watchService is closed,
+        stopped = true;
+        IOUtils.closeQuietly(watchService);
+
+        //Use atomic operations to avoid unnecessary exceptions caused by adding tasks to
+        //delayScheduler after delayScheduler is shutdown after stop
+        if (delayScheduler == null) {
+            return true;
+        }
+        atomicSchedulerOperation(() -> {
+            if (!delayScheduler.isShutdown()) {
+                delayScheduler.shutdownNow();
+            }
+        });
+        return delayScheduler.awaitTermination(timeout, unit);
+    }
+
+    private void watch() {
+        while (!stopped) {
+            try {
+                doWatch();
+            } catch (Throwable e) {
+                LOGGER.error("Error occur when watch {}.", path, e);
+            }
+        }
+    }
+
+    private void doWatch() {
+        WatchKey wk;
+        try {
+            wk = watchService.take();
+        } catch (ClosedWatchServiceException | InterruptedException e) {
+            return;
+        }
+
+        for (WatchEvent<?> event : wk.pollEvents()) {
+            File file = getFile(event, wk);
+            if (file == null) {
+                continue;
+            }
+            final WatchEvent.Kind<?> kind = event.kind();
+            if (kind == StandardWatchEventKinds.ENTRY_CREATE) {
+                pushEvent(new WatchEventContextImpl<>(event, file), create);
+            } else if (kind == StandardWatchEventKinds.ENTRY_MODIFY) {
+                pushEvent(new WatchEventContextImpl<>(event, file), modify);
+            } else if (kind == StandardWatchEventKinds.ENTRY_DELETE) {
+                pushEvent(new WatchEventContextImpl<>(event, file), delete);
+            } else if (kind == StandardWatchEventKinds.OVERFLOW) {
+                pushEvent(new WatchEventContextImpl<>(event, file), overflow);
+            }
+        }
+        wk.reset();
+    }
+
+    abstract void initDir(Path path);
+
+    abstract void register(Path path);
+
+    abstract File getFile(WatchEvent<?> event, WatchKey wk);
+
+    private void pushEvent(WatchEventContext<?> ctx, Consumer<WatchEventContext<?>> consumer) {
+        if (delay <= 0) {
+            consumer.accept(ctx);
+            return;
+        }
+
+        final String eventKey = ctx.file().getAbsolutePath() + ctx.event().kind().name();
+        if (eventSet.contains(eventKey)) {
+            return;
```

Comment:
can this operation make the event lose

Reply:
What the users mainly concerned is the type of event